File: src/main/java/no/novari/flyt/egrunnerverv/gateway/instance/mapping/EgrunnervervSakInstanceMappingService.java
package no.novari.flyt.egrunnerverv.gateway.instance.mapping;

import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.egrunnerverv.gateway.exception.ArchiveResourceNotFoundException;
import no.novari.flyt.egrunnerverv.gateway.exception.NonMatchingEmailDomainWithOrgIdException;
import no.novari.flyt.egrunnerverv.gateway.instance.ResourceRepository;
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakInstance;
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakKlassering;
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSaksPart;
import no.novari.flyt.egrunnerverv.gateway.slack.SlackAlertService;
import no.novari.flyt.instance.gateway.InstanceMapper;
import no.novari.flyt.instance.gateway.model.File;
import no.novari.flyt.instance.gateway.model.InstanceObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.function.Function;

@Service
@Slf4j
public class EgrunnervervSakInstanceMappingService implements InstanceMapper<EgrunnervervSakInstance> {

    @Value("${novari.flyt.egrunnerverv.checkSaksansvarligEpost:true}")
    boolean checkSaksansvarligEpost;

    @Value("${novari.flyt.egrunnerverv.checkEmailDomain:true}")
    boolean checkEmailDomain;
    private final ResourceRepository resourceRepository;

    @Value("${fint.org-id}")
    private String orgId;

    private final FormattingUtilsService formattingUtilsService;

    private final SlackAlertService slackAlertService;

    public EgrunnervervSakInstanceMappingService(
            ResourceRepository resourceRepository,
            FormattingUtilsService formattingUtilsService,
            SlackAlertService slackAlertService
    ) {
        this.resourceRepository = resourceRepository;
        this.formattingUtilsService = formattingUtilsService;
        this.slackAlertService = slackAlertService;
    }

    @Override
    public Mono<InstanceObject> map(
            Long sourceApplicationId,
            EgrunnervervSakInstance egrunnervervSakInstance,
            Function<File, Mono<UUID>> persistFile
    ) {
        return Mono.defer(() -> {

            String saksansvarligEpostFormatted = formattingUtilsService.formatEmail(egrunnervervSakInstance.getSaksansvarligEpost());

            if (checkEmailDomain) {
                String domain = formattingUtilsService.extractEmailDomain(saksansvarligEpostFormatted);
                if (!domain.equals(orgId)) {
                    throw new NonMatchingEmailDomainWithOrgIdException(domain, orgId);
                }
            }

            String saksansvarlig = "";

            if (checkSaksansvarligEpost) {
                saksansvarlig = resourceRepository
                        .getArkivressursHrefFromPersonEmail(saksansvarligEpostFormatted)
                        .orElseThrow(() -> new ArchiveResourceNotFoundException(
                                        saksansvarligEpostFormatted,
                                        slackAlertService
                                )
                        );
            }

            Map<String, String> valuePerKey = new HashMap<>();
            valuePerKey.put("sys_id", egrunnervervSakInstance.getSysId());
            valuePerKey.put("knr", egrunnervervSakInstance.getKnr());
            valuePerKey.put("gnr", egrunnervervSakInstance.getGnr());
            valuePerKey.put("bnr", egrunnervervSakInstance.getBnr());
            valuePerKey.put("fnr", egrunnervervSakInstance.getFnr());
            valuePerKey.put("snr", egrunnervervSakInstance.getSnr());
            valuePerKey.put("takstnummer", egrunnervervSakInstance.getTakstnummer());
            valuePerKey.put("tittel", egrunnervervSakInstance.getTittel());
            valuePerKey.put("saksansvarligEpost", saksansvarligEpostFormatted);
            valuePerKey.put("saksansvarlig", saksansvarlig);
            valuePerKey.put("eierforholdsnavn", egrunnervervSakInstance.getEierforholdsnavn());
            valuePerKey.put("eierforholdskode", egrunnervervSakInstance.getEierforholdskode());
            valuePerKey.put("prosjektnr", egrunnervervSakInstance.getProsjektnr());
            valuePerKey.put("prosjektnavn", egrunnervervSakInstance.getProsjektnavn());
            valuePerKey.put("kommunenavn", formattingUtilsService.formatKommunenavn(egrunnervervSakInstance.getKommunenavn()));
            valuePerKey.put("adresse", egrunnervervSakInstance.getAdresse());

            return Mono.just(
                    InstanceObject.builder()
                            .valuePerKey(valuePerKey)
                            .objectCollectionPerKey(Map.of(
                                    "saksparter", egrunnervervSakInstance.getSaksparter()
                                            .stream()
                                            .map(this::toInstanceObject)
                                            .toList(),
                                    "klasseringer", egrunnervervSakInstance.getKlasseringer()
                                            .stream()
                                            .map(this::toInstanceObject)
                                            .toList()
                            ))
                            .build()
            );
        });
    }

    private InstanceObject toInstanceObject(EgrunnervervSaksPart egrunnervervSaksPart) {
        return InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "navn", egrunnervervSaksPart.getNavn(),
                        "organisasjonsnummer", egrunnervervSaksPart.getOrganisasjonsnummer(),
                        "epost", egrunnervervSaksPart.getEpost(),
                        "telefon", egrunnervervSaksPart.getTelefon(),
                        "postadresse", egrunnervervSaksPart.getPostadresse(),
                        "postnummer", egrunnervervSaksPart.getPostnummer(),
                        "poststed", egrunnervervSaksPart.getPoststed()
                ))
                .build();
    }

    private InstanceObject toInstanceObject(EgrunnervervSakKlassering egrunnervervSakKlassering) {
        return InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "ordningsprinsipp", egrunnervervSakKlassering.getOrdningsprinsipp(),
                        "ordningsverdi", egrunnervervSakKlassering.getOrdningsverdi(),
                        "beskrivelse", egrunnervervSakKlassering.getBeskrivelse(),
                        "sortering", egrunnervervSakKlassering.getSortering(),
                        "untattOffentlighet", egrunnervervSakKlassering.getUntattOffentlighet()
                ))
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/instance/mapping/EgrunnervervSakInstanceMappingService.kt
package no.novari.flyt.egrunnerverv.gateway.instance.mapping

import no.novari.flyt.egrunnerverv.gateway.exception.ArchiveResourceNotFoundException
import no.novari.flyt.egrunnerverv.gateway.exception.NonMatchingEmailDomainWithOrgIdException
import no.novari.flyt.egrunnerverv.gateway.instance.ResourceRepository
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakInstance
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakKlassering
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSaksPart
import no.novari.flyt.egrunnerverv.gateway.slack.SlackAlertService
import no.novari.flyt.gateway.webinstance.InstanceMapper
import no.novari.flyt.gateway.webinstance.model.File
import no.novari.flyt.gateway.webinstance.model.instance.InstanceObject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class EgrunnervervSakInstanceMappingService(
    @param:Value("\${novari.flyt.egrunnerverv.checkSaksansvarligEpost:true}")
    private val checkSaksansvarligEpost: Boolean,
    @param:Value("\${novari.flyt.egrunnerverv.checkEmailDomain:true}")
    private val checkEmailDomain: Boolean,
    @param:Value("\${fint.org-id}")
    private val orgId: String,
    private val resourceRepository: ResourceRepository,
    private val formattingUtilsService: FormattingUtilsService,
    private val slackAlertService: SlackAlertService,
) : InstanceMapper<EgrunnervervSakInstance> {
    override fun map(
        sourceApplicationId: Long,
        incomingInstance: EgrunnervervSakInstance,
        persistFile: (File) -> UUID,
    ): InstanceObject {
        val saksansvarligEpostFormatted =
            formattingUtilsService.formatEmail(incomingInstance.saksansvarligEpost)

        if (checkEmailDomain) {
            val domain = formattingUtilsService.extractEmailDomain(saksansvarligEpostFormatted)
            if (domain != orgId) {
                throw NonMatchingEmailDomainWithOrgIdException(domain, orgId)
            }
        }

        val saksansvarlig =
            if (checkSaksansvarligEpost) {
                resourceRepository
                    .getArkivressursHrefFromPersonEmail(saksansvarligEpostFormatted)
                    ?: throw ArchiveResourceNotFoundException(
                        saksansvarligEpostFormatted,
                        slackAlertService,
                    )
            } else {
                ""
            }

        val valuePerKey =
            mutableMapOf(
                "sys_id" to incomingInstance.sysId,
                "knr" to incomingInstance.knr,
                "gnr" to incomingInstance.gnr,
                "bnr" to incomingInstance.bnr,
                "fnr" to incomingInstance.fnr,
                "snr" to incomingInstance.snr,
                "takstnummer" to incomingInstance.takstnummer,
                "tittel" to incomingInstance.tittel,
                "saksansvarligEpost" to saksansvarligEpostFormatted,
                "saksansvarlig" to saksansvarlig,
                "eierforholdsnavn" to incomingInstance.eierforholdsnavn,
                "eierforholdskode" to incomingInstance.eierforholdskode,
                "prosjektnr" to incomingInstance.prosjektnr,
                "prosjektnavn" to incomingInstance.prosjektnavn,
                "kommunenavn" to
                    formattingUtilsService
                        .formatKommunenavn(incomingInstance.kommunenavn)
                        .orEmpty(),
                "adresse" to incomingInstance.adresse,
            )

        return InstanceObject(
            valuePerKey,
            mutableMapOf(
                "saksparter" to incomingInstance.saksparter.map(this::toInstanceObject),
                "klasseringer" to incomingInstance.klasseringer.map(this::toInstanceObject),
            ),
        )
    }

    private fun toInstanceObject(egrunnervervSaksPart: EgrunnervervSaksPart): InstanceObject =
        InstanceObject(
            mapOf(
                "navn" to egrunnervervSaksPart.navn,
                "organisasjonsnummer" to egrunnervervSaksPart.organisasjonsnummer,
                "epost" to egrunnervervSaksPart.epost,
                "telefon" to egrunnervervSaksPart.telefon,
                "postadresse" to egrunnervervSaksPart.postadresse,
                "postnummer" to egrunnervervSaksPart.postnummer,
                "poststed" to egrunnervervSaksPart.poststed,
            ),
            mutableMapOf(),
        )

    private fun toInstanceObject(egrunnervervSakKlassering: EgrunnervervSakKlassering): InstanceObject =
        InstanceObject(
            mapOf(
                "ordningsprinsipp" to egrunnervervSakKlassering.ordningsprinsipp,
                "ordningsverdi" to egrunnervervSakKlassering.ordningsverdi,
                "beskrivelse" to egrunnervervSakKlassering.beskrivelse,
                "sortering" to egrunnervervSakKlassering.sortering,
                "untattOffentlighet" to egrunnervervSakKlassering.untattOffentlighet,
            ),
            mutableMapOf(),
        )
}
